src/BTVNanoCommissioning/workflows/BTA_ttbar_producer.py

```diff
@@ -0,0 +1,527 @@
+                abs(_fix(p.pdgId)) // 1000 == pid
+            )
+
+            sel = (
+                is_lep(events.GenPart)
+                & ((events.GenPart.statusFlags & (1 << 13)) > 0)
+                & (events.GenPart.pt > 3.0)
+            )  # requires pT > 3 GeV
+            genlep = events.GenPart[sel]
+
+            # trace parents up to 4 generations (from BTA code)
+            genlep_pa1G = genlep.parent
+            genlep_pa2G = genlep.parent.parent
+            genlep_pa3G = genlep.parent.parent.parent
+            genlep_pa4G = genlep.parent.parent.parent.parent
+            istau = genlep_pa1G.pdgId == 15
+            isWZ = is_WZ(genlep_pa1G) | is_WZ(genlep_pa2G)
+            isD = is_heavy_hadron(genlep_pa1G, 4) | is_heavy_hadron(genlep_pa2G, 4)
+            isB = (
+                is_heavy_hadron(genlep_pa1G, 5)
+                | is_heavy_hadron(genlep_pa2G, 5)
+                | is_heavy_hadron(genlep_pa3G, 5)
+                | is_heavy_hadron(genlep_pa4G, 5)
+            )
+
+            Genlep = ak.zip(
+                {
+                    "pT": genlep.pt,
+                    "eta": genlep.eta,
+                    "phi": genlep.phi,
+                    "pdgID": genlep.pdgId,
+                    "status": genlep.status,
+                    "mother": ak.fill_none(
+                        ak.where(isB | isD, 5 * isB + 4 * isD, 10 * istau + 100 * isWZ),
+                        0,
+                    ),
+                }
+            )
+
+            ###################
+            #  Gen particles  #
+            ###################
+            # Gen particles of interest in the BTA ttbar workflow
+            genpart = events.GenPart[
+                (events.GenPart.statusFlags & (1 << 7)) > 0
+            ]  # isHardProcess
+            gen_channel = ak.prod(
+                genpart[(abs(genpart.pdgId) == 11) | (abs(genpart.pdgId) == 13)].pdgId,
+                axis=1,
+            )
+
+            gen_arrays = ak.zip(
+                {
+                    "pt": genpart.pt,
+                    "eta": genpart.eta,
+                    "phi": genpart.phi,
+                    "m": genpart.mass,
+                    "id": genpart.pdgId,
+                }
+            )
+
+            #################
+            #  Gen weights  #
+            #################
+            # top pT reweighting, according to BTA code
+            genttbar = events.GenPart[
+                ((events.GenPart.statusFlags & (1 << 13)) > 0)
+                & (abs(events.GenPart.pdgId) == 6)
+            ]  # is last copy and is t/tbar
+            gentop = ak.firsts(genttbar[genttbar.pdgId == 6])
+            genantitop = ak.firsts(genttbar[genttbar.pdgId == -6])
+            basic_vars["ttbar_ptweight"] = ak.fill_none(
+                np.sqrt(
+                    np.exp(0.0615 - 0.0005 * gentop.pt)
+                    * np.exp(0.0615 - 0.0005 * genantitop.pt)
```

Comment:
is the top pt reweighting function still up-to-date? there was an update during Run 2 (mostly because the old formula considered only tops up to 500 GeV, the new one considers now up to 1 TeV), see details [here](https://twiki.cern.ch/twiki/bin/view/CMS/TopPtReweighting)

Reply:
I think there is also a special treatment for the > 500 GeV bin IIRC, please check

Reply:
It's actually included in the framework already 😉 https://github.com/cms-btv-pog/BTVNanoCommissioning/blob/master/src/BTVNanoCommissioning/utils/correction.py#L1257-L1273

you can just input the `GenPart` to the topreweight function, it should work out

Reply:
You are right, indeed in BTV we should apply the data-NLO one(the NNLO-NLO is mostly used in the analysis)

Then we should still use the formula there. With pT>500GeV we use the value at 500GeV. and seems you are taking the NNLO-data version I think should be fine

Reply:
ah thanks! But the twiki seems to suggest the >500 issue should apply in data-NLO. I'll change the code accordingly